x-pack/platform/packages/shared/kbn-classic-stream-flyout/src/components/stream_name_input/stream_name_input.tsx

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useState, useMemo, useCallback, useEffect } from 'react';
+import { EuiFieldText, EuiFlexGroup, EuiFlexItem, useEuiTheme } from '@elastic/eui';
+import { css } from '@emotion/react';
+import {
+  parseIndexPattern,
+  createInputGroups,
+  countWildcards,
+  buildStreamName,
+  type ValidationErrorType,
+} from '../../utils';
+
+export interface StreamNameInputProps {
+  /** The index pattern containing wildcards (e.g., "*-logs-*-*") */
+  indexPattern: string;
+  /** Callback when the stream name changes */
+  onChange: (streamName: string) => void;
+  /**
+   * Validation error type. When 'empty', only empty inputs are highlighted.
+   * For other error types, all inputs are highlighted.
+   */
+  validationError?: ValidationErrorType;
+  /** Test subject prefix for the inputs */
+  'data-test-subj'?: string;
+}
+
+export const StreamNameInput = ({
+  indexPattern,
+  onChange,
+  validationError = null,
+  'data-test-subj': dataTestSubj = 'streamNameInput',
+}: StreamNameInputProps) => {
+  const { euiTheme } = useEuiTheme();
+
+  const segments = useMemo(() => parseIndexPattern(indexPattern), [indexPattern]);
+  const inputGroups = useMemo(() => createInputGroups(segments), [segments]);
+  const wildcardCount = useMemo(() => countWildcards(indexPattern), [indexPattern]);
+
+  // Internal state for wildcard values
+  const [parts, setParts] = useState<string[]>(() => Array(wildcardCount).fill(''));
+
+  // Reset parts when index pattern changes
+  useEffect(() => {
+    setParts(Array(wildcardCount).fill(''));
+  }, [indexPattern, wildcardCount]);
+
+  // Update stream name when wildcard values change
+  useEffect(() => {
+    const streamName = buildStreamName(indexPattern, parts);
+    onChange(streamName);
+  }, [indexPattern, parts, onChange]);
```

Comment:
both of these can be dissolved since `useEffect()` should only be used in very specific narrow set of use-cases and should not be used in 98% of cases well outlined here https://react.dev/learn/you-might-not-need-an-effect

There a couple of options: lift state up or use key={indexPattern} from parentComponent to reset identity of component

streamName is a typical derived state that can live at the top of function body and onChange can be called conditionally when `indexPattern` changes inline in the body as they recommend it in the react docs.

In my rewrite suggestion https://github.com/damian-polewski/kibana/pull/2/files I'm actually rolling with lifting state up. But see for yourself which one you like more